metrics_collector/snapshot_collector.go

```diff
@@ -0,0 +1,93 @@
+package metricscollector
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+)
+
+type SnapshotCollector struct {
+	*baseCollector
+
+	sizeTotalMetric metricInfo
+}
+
+func NewSnapshotCollector(
+	logger logrus.FieldLogger,
+	nodeID string,
+	ds *datastore.DataStore) *SnapshotCollector {
+
+	snapshotCollector := &SnapshotCollector{
+		baseCollector: newBaseCollector(subsystemSnapshot, logger, nodeID, ds),
+	}
+
+	snapshotCollector.sizeTotalMetric = metricInfo{
+		Desc: prometheus.NewDesc(
+			prometheus.BuildFQName(longhornName, subsystemSnapshot, "size_total_bytes"),
+			"Sum of size of snapshots owned by this node",
+			[]string{countLabel, userCreatedLabel},
+			nil,
+		),
+		Type: prometheus.GaugeValue,
+	}
+
+	return snapshotCollector
+}
+
+func (c *SnapshotCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.sizeTotalMetric.Desc
+}
+
+func (c *SnapshotCollector) Collect(ch chan<- prometheus.Metric) {
+	defer func() {
+		if err := recover(); err != nil {
+			c.logger.WithField("error", err).Warn("Panic during collecting metrics")
+		}
+	}()
+
+	snapshotLists, err := c.ds.ListSnapshotsRO(labels.Everything())
+	if err != nil {
+		c.logger.WithError(err).Warn("Error during scrape")
+		return
+	}
+
+	snapshotSizeTotal := struct {
+		user, system struct {
+			size  int64
+			count int
+		}
+	}{}
+
+	for _, snapshot := range snapshotLists {
+		volume, err := c.ds.GetVolumeRO(snapshot.Spec.Volume)
+		if err != nil {
+			logrus.WithError(err).Warnf("failed to get volume %v for snapshot %v", snapshot.Spec.Volume, snapshot.Name)
+			continue
+		}
+
+		if volume.Status.OwnerID != c.currentNodeID {
+			continue
+		}
+
+		// Skip volume-head because it is not a real snapshot.
+		if snapshot.Name == "volume-head" {
+			continue
+		}
+
+		if snapshot.Status.UserCreated {
+			snapshotSizeTotal.user.size += snapshot.Status.Size
+			snapshotSizeTotal.user.count++
+		} else {
+			snapshotSizeTotal.system.size += snapshot.Status.Size
+			snapshotSizeTotal.system.count++
+		}
+	}
+
+	ch <- prometheus.MustNewConstMetric(c.sizeTotalMetric.Desc, c.sizeTotalMetric.Type, float64(snapshotSizeTotal.user.size), strconv.Itoa(snapshotSizeTotal.user.count), "true")
+	ch <- prometheus.MustNewConstMetric(c.sizeTotalMetric.Desc, c.sizeTotalMetric.Type, float64(snapshotSizeTotal.system.size), strconv.Itoa(snapshotSizeTotal.system.count), "false")
+}
```

Comment:
I think that we don't need to sump up the size in longhorn-manager because user can write Prometheus query to sume the data later. Proposing the metric definition like:
```go


type SnapshotCollector struct {
	*baseCollector

	sizeMetric metricInfo
}

func NewSnapshotCollector(
	logger logrus.FieldLogger,
	nodeID string,
	ds *datastore.DataStore) *SnapshotCollector {

	snapshotCollector := &SnapshotCollector{
		baseCollector: newBaseCollector(subsystemSnapshot, logger, nodeID, ds),
	}

	snapshotCollector.sizeMetric = metricInfo{
		Desc: prometheus.NewDesc(
			prometheus.BuildFQName(longhornName, subsystemSnapshot, "size_bytes"),
			"Actual space used by the snapshot",
			[]string{nodeLabel, volumeLabel, userCreatedLabel},
			nil,
		),
		Type: prometheus.GaugeValue,
	}

	return snapshotCollector
}


```